Provide property tests for `changeUTxO`

jonathanknowles opened the pull request:
### Issue Number

ADP-1084

### Comments

This PR provides property tests for the `changeUTxO` function.

Since this function is re-used by `totalUTxO` (which we want to modify and then verify), it's useful to first verify that `changeUTxO` is correct.

jonathanknowles committed:
Add `null` and `size` operations to `UTxO` type.

These operations just reuse the underlying operations provided by the
`Map` module.

Exporting these operations means that we don't have to unwrap UTxO
values in order to determine their size.

jonathanknowles committed:
Add property tests for `changeUTxO` function.

rvl commented on lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -425,6 +438,202 @@ prop_availableUTxO makeProperty =
+    forAllShrink (listOf genTxOutputs) (shrinkList shrinkTxOutputs) inner
+  where
+    inner :: [TxOutputs] -> Property
+    inner txOutputs =
```
```suggestion
newtype Classify a = Classify (a -> Bool)

prop_changeUTxO :: Classify Address -> [TxOut] -> Property
prop_changeUTxO (Classify condition) txOutputs =
```

jonathanknowles replied:
@rvl Nice suggestion!

I've actually made a similar change in 0ffa43da16b9232aafcf613453709a279545cc74, which replaces the various types with a single `IsOursIf {condition :: a -> Bool}` newtype (with a single instance).

rvl commented on lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -425,6 +438,202 @@ prop_availableUTxO makeProperty =
+
+-- | Represents all the outputs of a transaction (both change and non-change).
+--
+newtype TxOutputs = TxOutputs
```
I don't get why we go to the trouble of making a `newtype`, but don't add an `Arbitrary` instance, and then use `shrinkAll`, when a simple `[TxOut]` parameter on the property would do the same thing.

jonathanknowles replied:
@rvl Fair enough!

I've added an `Arbitrary` instance for `TxOutputs` in 761ecc21a740e95ef5367a9a41919edfbbf7763a.

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -425,6 +438,202 @@ prop_availableUTxO makeProperty =
+-- | An implementation of 'IsOurs' that matches all addresses.
+--
+instance IsOurs All Address where
+    isOurs = isOursIf (const True)
```
**Suggestion**: replace each of the type class instances with a single instance that encapsulates a match condition.

jonathanknowles replied:
Fixed in 0ffa43da16b9232aafcf613453709a279545cc74.

rvl commented on lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -425,6 +438,202 @@ prop_availableUTxO makeProperty =
+--    - 0b11111111 : even (Hamming weight = 8)
+--
+addressParity :: Address -> Parity
+addressParity = parity . addressPopCount
```
Any other `Address -> Bool` function would be just as good, as long it splits the set of generated addresses into roughly equal parts.

jonathanknowles replied:
Yes, definitely.

It's an arbitrary choice (excuse the pun!) to use this particular function. It has two nice advantages:

1. Addresses are "even" or "odd" with equal frequency (meaning we don't have to do anything special with generation).
2. Address parity can be determined with a pure function (meaning we don't have to keep any state around in our `IsOurs` instance).

I've added a comment to `prop_changeUTxO` with these reasons.

jonathanknowles committed:
Simplify `IsOurs` type class instances for `changeUTxO` property tests.

In response to:

https://github.com/input-output-hk/cardano-wallet/pull/2836/files#r694419473

rvl commented on lib/core/src/Cardano/Wallet/Primitive/Types/UTxO.hs:
```diff
@@ -127,6 +130,12 @@ restrictedTo :: UTxO -> Set TxOut -> UTxO
     UTxO $ Map.filter (`Set.member` outs) utxo
 
+null :: UTxO -> Bool
+null (UTxO u) = Map.null u
```
Really? The only place this is used it's with `(not (UTxO.null x)`. But `x /= mempty` would be fine.

jonathanknowles replied:
> Really? The only place this is used it's with `(not (UTxO.null x)`. But `x /= mempty` would be fine.

I agree with you that `x /= mempty` would work fine!

Though I find that `UTxO.null` is a little more self-documenting (in this case), as:

- A reader doesn't have to navigate to the definition to figure out the type of thing being tested, because it's stated right at the point of use; and
- The `null` function is idiomatic across different container types, so it's familiar.

jonathanknowles committed:
Make `Arbitrary` instance for `TxOutputs`.

In response to:

https://github.com/input-output-hk/cardano-wallet/pull/2836#discussion_r694417920

Anviking committed:
Fix shellcheck warnings in ./scripts/make_release.sh

Since #2840 was merged manually, this wasn't caught there.

I re-tested the command, and it seems to still work.
